File: back-end/community/Subscriber.py
from web3 import Web3
# from ??? import USER_ADDRESS
from ..contracts.Contracts import subscribers_contract
from ..exceptions.AlreadySubscribeExeption import AlreadySubscribeExeption
from ..exceptions.AlreadyUnsubscribeExeption import AlreadyUnsubscribeExeption
from ..exceptions.SubscribeToYourself import SubscribeToYourself
from ..exceptions.UnsubscribeFromYourself import UnsubscribeFromYourself


class Subscriber():
    def subscribe_on(self, account: str):
        if (account == USER_ADDRESS):
            raise SubscribeToYourself
        transaction = subscribers_contract.functions.subscribeOn(account).transact()
        receipt = Web3.eth.wait_for_transaction_receipt(transaction)
        if (receipt == 0):
            raise AlreadySubscribeExeption

    def unsubscribe_from(self, account: str):
        if (account == USER_ADDRESS):
            raise UnsubscribeFromYourself
        transaction = subscribers_contract.functions.unsubscribeFrom(account).transact()
        receipt = Web3.eth.wait_for_transaction_receipt(transaction)
        if (receipt == 0):
            raise AlreadyUnsubscribeExeption

    def get_subscription_count(self, account: str) -> int:
        return subscribers_contract.functions.getSubscribersAmount(account).call()

File: back-end/community/Subscriber.js
class Subscriber {
    async subscribe_on(account) {
        if (await getCookie('account') === account)
            throw 'Subscribe to yourself'
        let subscribe_dict_transaction = {
            'value': 0,
            'gas': 100000,
            'chainId': await web3.eth.getChainId().then(count => parseInt(count, 10)),
            'to': subscribers_address,
            'from': await getCookie('account'),
            'gas_price': await web3.eth.getGasPrice().then(count => parseInt(count, 10)),
            'nonce': await web3.eth.getTransactionCount(await getCookie('account'), 'pending').then(count => parseInt(count, 10)),
        }
        await subscribers_contract.methods.subscribeOn(account).send(subscribe_dict_transaction)
    }

    async unsubscribe_from(account) {
        if (await getCookie('account') === account_)
            throw 'Unsubscribe from yourself'
        let unsubscribe_dict_transaction = {
            'value': 0,
            'gas': 100000,
            'chainId': await web3.eth.getChainId().then(count => parseInt(count, 10)),
            'to': subscribers_address,
            'from': await getCookie('account'),
            'gas_price': await web3.eth.getGasPrice().then(count => parseInt(count, 10)),
            'nonce': await web3.eth.getTransactionCount(await getCookie('account'), 'pending').then(count => parseInt(count, 10)),
        }
        await subscribers_contract.methods.unsubscribeFrom(account).send(unsubscribe_dict_transaction)
    }
    get_subscription_count(account) {
        return subscribers_contract.methods.getSubscribersAmount(account).call()
    }
}
